crates/net/eth-wire-types/src/eth_snap.rs

```diff
@@ -0,0 +1,278 @@
+    #[inline]
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
+}
+
+impl<S, E, N> Stream for EthSnapStream<S, N>
+where
+    S: Stream<Item = Result<BytesMut, E>> + Unpin,
+    EthSnapStreamError: From<E>,
+    N: NetworkPrimitives,
+{
+    type Item = Result<EthSnapMessage<N>, EthSnapStreamError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        let res = ready!(this.inner.poll_next(cx));
+
+        match res {
+            Some(Ok(bytes)) => Poll::Ready(Some(this.eth_snap.decode_message(bytes))),
+            Some(Err(err)) => Poll::Ready(Some(Err(err.into()))),
+            None => Poll::Ready(None),
+        }
+    }
+}
+
+impl<S, E, N> Sink<EthSnapMessage<N>> for EthSnapStream<S, N>
+where
+    S: Sink<Bytes, Error = E> + Unpin,
+    EthSnapStreamError: From<E>,
+    N: NetworkPrimitives,
+{
+    type Error = EthSnapStreamError;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.project().inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: EthSnapMessage<N>) -> Result<(), Self::Error> {
+        let mut this = self.as_mut().project();
+
+        let bytes = match item {
+            EthSnapMessage::Eth(eth_msg) => this.eth_snap.encode_eth_message(eth_msg)?,
+            EthSnapMessage::Snap(snap_msg) => this.eth_snap.encode_snap_message(snap_msg),
+        };
+
+        this.inner.start_send_unpin(bytes)?;
+        Ok(())
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.project().inner.poll_flush(cx).map_err(Into::into)
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.project().inner.poll_close(cx).map_err(Into::into)
+    }
+}
+
+/// Stream Handling combined ETH+SNAP protocol logic
+#[derive(Debug, Clone)]
+struct EthSnapStreamInner<N> {
+    /// ETH protocol version
+    eth_version: EthVersion,
+    /// Type marker
+    _pd: PhantomData<N>,
+}
+
+impl<N> EthSnapStreamInner<N>
+where
+    N: NetworkPrimitives,
+{
+    /// Create a new ETH+SNAP protocol stream
+    const fn new(eth_version: EthVersion) -> Self {
+        Self { eth_version, _pd: PhantomData }
+    }
+
+    #[inline]
+    const fn eth_version(&self) -> EthVersion {
+        self.eth_version
+    }
+
+    fn decode_message(&self, bytes: BytesMut) -> Result<EthSnapMessage<N>, EthSnapStreamError> {
+        if bytes.len() > MAX_MESSAGE_SIZE {
+            return Err(EthSnapStreamError::MessageTooLarge(bytes.len(), MAX_MESSAGE_SIZE));
+        }
+
+        if bytes.is_empty() {
+            return Err(EthSnapStreamError::Rlp(alloy_rlp::Error::InputTooShort));
+        }
+
+        let message_id = bytes[0];
+
+        if message_id <= crate::message::EthMessageID::max() {
```

Comment:
this works, but only because of:

https://github.com/paradigmxyz/reth/blob/54b0d357f987f9d1e973a447b815076cef520823/crates/net/eth-wire/src/capability.rs#L272-L283

let's add a note here that this check here is sound